JavaScript/JavaScriptSDK.Tests/Selenium/ai.tests.d.ts

```diff
@@ -1708,12 +1728,16 @@ declare module Microsoft.ApplicationInsights {
          *
          * Note: XDomainRequest does not support sync requests. This 'isAsync' parameter is added
          * to maintain consistency with the xhrSender's contract
+         * Note: XDomainRequest does not support custom headers and we are not able to get
+         * appId from the backend for the correct correlation.
          */
         private _xdrSender(payload, isAsync);
         /**
          * Send Beacon API request
          * @param payload {string} - The data payload to be sent.
          * @param isAsync {boolean} - not used
+         * Note: Beacon API does not support custom headers and we are not able to get
```

Comment:
Why was the design to go with AppId and not Ikey? Though Ikey is mutable for a resource, it still has to be valid in order to send data to. Seems beacon api users wont get the benefit of this feature, and going with ikey would have been useful here?

Reply:
@SergeyKanzhelev, @jpiyali,
Correct me if I'm wrong - are we talking about the following?
1) pass ikey:<ikey> in a Request-Context header 
2) add <ikey> to the source of the server-side request
3) modify UX to work with <ikey> from source

It'll be very difficult to change UX to work with both ikeys and appIds as all Draft queries are bound only to appId and it cannot be implemented without significant refactoring and performance degradation. 

We can proceed in two ways:
1) Ensure that current approach works and then remove custom header requirement from Breeze - actually nothing will change because Breeze responds with appId today only if batch from the same key
2) Support ikey:<ikey> format but extend server SDK, to resolve it to appId and append appId to the request source like it's done today.

Reply:
I wouldn't make it server SDK responsibility. Otherwise it will create another attack vector - sending many different ikeys will force an app to run many queries to Breeze. Plus there will be a delay in obtaining this mapping that will delay app delivery.

Does Kusto supports addressing an application by ikey? I know it supports by name and ARM ID. Is it slower request? I don't understand why performance will suffer - it is basically one indirection in the query preparation. Actual query will be the same.

Reply:
No, Kusto doesn't support ikey. We solved this problem before once - when we sent hashed ikey - and it resolved in a heavy queries for finding right appId or componentId by ikey. Even if we'll send ikey (instead of hashed one - and that still slightly concerns me from the security point) then we'll need to resolve it to appId anyway by additional call to Breeze or CDS in order to get information from Draft. Our UX queries are complicated and slow at the moment and we keep constantly fighting with the performance overhead and ongoing optimization. 

So my proposal here would be not to support Beacon API now as it's disabled by the default and is used only by couple customers who are dealing with other problems with JS SDK + Beacon API (basically it doesn't work)... And later on, when we'll completely test the new channel and ensure that it's working, remove Sdk-Context header requirement from Breeze and return appId in any case. BTW this header was introduced to limit issues with multi-ikey scenarios, but we found other security issues and now Breeze return appId only if all items from the batch are from the same ikey. In this logic it's safe to remove header.

Reply:
Thank you for the explanation. Perhaps the feature that allows to pre-set the `ikey` to `appid` mapping may be useful here. Basically let customer configure it in application deployment time. On certain environments like web apps we can pre-set it during the deployment

Reply:
this feature will address beacon API support concern

Reply:
@SergeyKanzhelev, please check out my latest commit with exposing appId in the config.